rust/src/applayerpgsql/parser.rs

```diff
@@ -0,0 +1,597 @@
+        }
+    }
+}
+
+impl From<u8> for PgsqlErrorNoticeFieldTypes {
+    fn from(identifier: u8) -> PgsqlErrorNoticeFieldTypes {
+        match identifier {
+            0x53 => PgsqlErrorNoticeFieldTypes::SeverityLocalizable,
+            0x56 => PgsqlErrorNoticeFieldTypes::SeverityNonLocalizable,
+            0x43 => PgsqlErrorNoticeFieldTypes::CodeSqlStateCode,
+            0x4d => PgsqlErrorNoticeFieldTypes::Message,
+            0x44 => PgsqlErrorNoticeFieldTypes::Detail,
+            0x48 => PgsqlErrorNoticeFieldTypes::Hint,
+            0x50 => PgsqlErrorNoticeFieldTypes::Position,
+            0x70 => PgsqlErrorNoticeFieldTypes::InternalPosition,
+            0x71 => PgsqlErrorNoticeFieldTypes::InternalQuery,
+            0x57 => PgsqlErrorNoticeFieldTypes::Where,
+            0x73 => PgsqlErrorNoticeFieldTypes::SchemaName,
+            0x74 => PgsqlErrorNoticeFieldTypes::TableName,
+            0x63 => PgsqlErrorNoticeFieldTypes::ColumnName,
+            0x64 => PgsqlErrorNoticeFieldTypes::DataType,
+            0x6e => PgsqlErrorNoticeFieldTypes::ConstraintName,
+            0x46 => PgsqlErrorNoticeFieldTypes::File,
+            0x4c => PgsqlErrorNoticeFieldTypes::Line,
+            0x52 => PgsqlErrorNoticeFieldTypes::Routine,
+            0x00 => PgsqlErrorNoticeFieldTypes::TerminatorToken,
+            _ => PgsqlErrorNoticeFieldTypes::UnknownFieldType, // adding this because documentation says "rontends should silently ignore fields of unrecognized type."
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct PgsqlErrorNoticeMessageField<'a>{
+    field_type: PgsqlErrorNoticeFieldTypes, // TODO should this become an enum?
+    field_value: Option<&'a[u8]>, // TODO Would it be better if this were a String or a &str?
+}
+
+
+// Parse an error response with non-localizeable severity message.
+// Possible values: ERROR, FATAL, or PANIC
+named!(pub parse_error_response_severity<PgsqlErrorNoticeMessageField>,
+    do_parse!(
+        field_type: char!('V')
+        >> field_value: alt!(tag!("ERROR") | tag!("FATAL") | tag!("PANIC"))
+        >> tag!("\x00")
+        >> (PgsqlErrorNoticeMessageField{
+                field_type: PgsqlErrorNoticeFieldTypes::from(field_type),
+                field_value: Some(field_value)
+        })
+    ));
+
+named!(pub parse_error_response_field<PgsqlErrorNoticeMessageField>,
+    switch!(peek!(be_u8),
+        b'\0' => do_parse!(
+            field_type: be_u8
+            >> data: eof!()
+            >> (PgsqlErrorNoticeMessageField{
+                field_type: PgsqlErrorNoticeFieldTypes::from(field_type),
+                field_value: None
+            })
+        ) |
+        _ => do_parse!(
+            field_type: be_u8 
+            >> field_value: opt!(take_until!("\x00"))
+            >> tag!("\x00")
+            >> (PgsqlErrorNoticeMessageField{
+                field_type: field_type.into(), // the issue with this approach is that we accept invalid identifiers...
```

Comment:
In the `switch!` case you should be able to handle all the valid ones, right?

Instead of `_ => do_parse!(`, something like
```rust
b'A' | b'B' | b'C' => do_parse!(
```